### PR TITLE
native: unify and simplify usage of tapsetup script

### PR DESCRIPTION
**Pull request by authmillenon**
Started with the intention to make the deactivation of IPv6 optional, ended up overhauling the whole script(s).

From the self-docs:

```
usage: tapsetup [OPTIONS]

  If no option is given, -c is assumed

  OPTIONS:
  -c [<num>], --create [<num>]: Create <num> tap interfaces (default: 2)
                  -d, --delete: Delete all interface
    -b <name>, --bridge <name>: Give name for the bridge (default: tapbr)
       -t <name>, --tap <name>: name base for the tap interfaces; the
                                generated names will be <name>x
                                (default: tap; ignored on OSX and FreeBSD)
         -6, --deactivate-ipv6: deactivate IPv6 for the interfaces and bridge
                                (ignored on OSX and FreeBSD)
                    -h, --help: prints this text
```

### CODE REVIEW

**Comment by authmillenon** [comment]
Should work for Linux, Mac OS X, and FreeBSD. Please test.

File: cpu/native/tapsetup [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+
+PROGRAM=$(basename $0)
+COUNT="2"
+COMMAND=""
+BRNAME="tapbr0"
+TAPNAME="tap"
+DEACTIVATE_IPV6=""
+
+usage() {
+    echo "usage: $PROGRAM [OPTIONS]" >&2
```
Comment (LudwigOrtmann) [inline cpu/native/tapsetup]: Please surround all variable names with curly braces, i.e. `$PROGRAM` -> `${PROGRAM}`.

Comment (authmillenon) [inline cpu/native/tapsetup]: Will do. Any particular reason why?

Comment (authmillenon) [inline cpu/native/tapsetup]: Addressed

**Comment by authmillenon** [comment]
(this removes the script's dependency on `brutils` BTW, since it uses `iproute2` on Linux for bridge configuration instead of it).

File: cpu/native/tapsetup [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+    echo "  OPTIONS:" >&2
+    echo "  -c [<num>], --create [<num>]: Create <num> tap interfaces (default: 2)" >&2
+    echo "                  -d, --delete: Delete all interface" >&2
+    echo "    -b <name>, --bridge <name>: Give name for the bridge (default: tapbr)" >&2
+    echo "       -t <name>, --tap <name>: name base for the tap interfaces; the" >&2
+    echo "                                generated names will be <name>x" >&2
+    echo "                                (default: tap; ignored on OSX and FreeBSD)" >&2
+    echo "         -6, --deactivate-ipv6: deactivate IPv6 for the interfaces and bridge" >&2
+    echo "                                (ignored on OSX and FreeBSD)" >&2
+    echo "                    -h, --help: prints this text" >&2
```
Comment (LudwigOrtmann) [inline cpu/native/tapsetup]: please make the help options left-aligned

File: cpu/native/tapsetup [hunk 3]
```diff
@@ -0,0 +1,211 @@
+#!/bin/sh
+
+PROGRAM=$(basename $0)
+COUNT="2"
+COMMAND=""
+BRNAME="tapbr0"
```
Comment (LudwigOrtmann) [inline cpu/native/tapsetup]: This default does not work on OSX:
```
creating tapbr0
ifconfig: SIOCIFCREATE2: Invalid argument
```

File: cpu/native/tapsetup [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+            case "$2" in
+                "")
+                    usage
+                    exit 2 ;;
+                *)
+                    BRNAME="$2"
+                    shift 2 ;;
+            esac ;;
+        -c|--create)
+            if [ COMMAND = "delete" ]; then
```
Comment (LudwigOrtmann) [inline cpu/native/tapsetup]: a check for a non-empty command would be safer

File: cpu/native/tapsetup (hunk 3 above)
Comment (authmillenon) [inline cpu/native/tapsetup]: Then what does?

File: cpu/native/tapsetup [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+                    shift ;;
+                *[!0-9]*:)
+                    usage
+                    exit 2;;
+                *)
+                    COUNT="$2"
+                    shift 2 ;;
+            esac ;;
+        -d|--delete)
+            if [ COMMAND = "create" ]; then
```
Comment (LudwigOrtmann) [inline cpu/native/tapsetup]: compare https://github.com/RIOT-OS/RIOT/pull/3182/files#r32370373

File: cpu/native/tapsetup (hunk 3 above)
Comment (LudwigOrtmann) [inline cpu/native/tapsetup]: I suggest you compare the script you replace ;)

Comment (LudwigOrtmann) [inline cpu/native/tapsetup]: (OSX insists on bridges with a 'bridge' prefix..)

Comment (LudwigOrtmann) [inline cpu/native/tapsetup]: same goes for FreeBSD

File: cpu/native/tapsetup [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+}
+
+create_tap() {
+    case "${PLATFORM}" in
+        FreeBSD)
+            echo "creating ${TAPNAME}${N}" || exit 1
+            sudo ifconfig tap${N} create || exit 1
+            sudo chown ${USER} /dev/tap${N} || exit 1
+            sudo ifconfig tap${N} up || exit 1
+            sudo ifconfig bridge0 addm tap${N} || exit 1 ;;
```
Comment (LudwigOrtmann) [inline cpu/native/tapsetup]: `bridge0` fail

**Comment by LudwigOrtmann** [comment]
Something fails on FreeBSD:
```
~/RIOT/examples/default/bin/native/default.elf: open(/dev/tap0): Operation not permitted
```

**Comment by LudwigOrtmann** [comment]
this PR:
```
+ basename ./cpu/native/tapsetup
+ PROGRAM=tapsetup
+ COUNT=2
+ COMMAND=''
+ BRNAME=tapbr0
+ TAPNAME=tap
+ DEACTIVATE_IPV6=''
+ true
+ BRNAME=bridge0
+ shift 2
+ true
+ break
+ [ -z lo ]
+ [ -z '' ]
+ COMMAND=create
+ uname -s
+ PLATFORM=FreeBSD
+ [ create = create ]
+ create_bridge
+ echo 'creating bridge0'
creating bridge0
+ sudo kldload if_bridge
+ sudo ifconfig bridge0 create
+ begin_tap
+ sudo kldload if_tap
+ seq 0 1
+ create_tap
+ echo 'creating tap0'
creating tap0
+ sudo ifconfig tap0 create
+ sudo chown lo /dev/tap0
+ sudo ifconfig tap0 up
+ sudo ifconfig bridge0 addm tap0
+ create_tap
+ echo 'creating tap1'
creating tap1
+ sudo ifconfig tap1 create
+ sudo chown lo /dev/tap1
+ sudo ifconfig tap1 up
+ sudo ifconfig bridge0 addm tap1
+ up_bridge
+ sudo ifconfig bridge0 up
```
```
+ basename ./cpu/native/tapsetup
+ PROGRAM=tapsetup
+ COUNT=2
+ COMMAND=''
+ BRNAME=tapbr0
+ TAPNAME=tap
+ DEACTIVATE_IPV6=''
+ true
+ BRNAME=bridge0
+ shift 2
+ true
+ [ COMMAND = create ]
+ COMMAND=delete
+ shift
+ true
+ break
+ [ -z lo ]
+ [ -z delete ]
+ uname -s
+ PLATFORM=FreeBSD
+ [ delete = create ]
+ [ delete = delete ]
+ delete_bridge
+ echo 'deleting bridge0'
deleting bridge0
+ sudo kldunload if_tap
+ sudo kldunload if_bridge
+ exit 0
```

master:
```
+ COMMAND=create
+ COUNT=''
+ DEFCOUNT=2
+ [ -z lo ]
+ [ -z create ]
+ [ create = create ]
+ [ -z '' ]
+ COUNT=2
+ sudo kldload if_tap
+ sudo kldload if_bridge
+ sudo sysctl net.link.tap.user_open=1
net.link.tap.user_open: 0 -> 1
+ echo 'creating  ...'
creating  ...
+ sudo ifconfig bridge0 create
+ sudo ifconfig bridge0 up
+ seq 0 1
+ sudo ifconfig tap0 create
+ sudo chown lo /dev/tap0
+ sudo ifconfig tap0 up
+ sudo ifconfig bridge0 addm tap0
+ sudo ifconfig tap1 create
+ sudo chown lo /dev/tap1
+ sudo ifconfig tap1 up
+ sudo ifconfig bridge0 addm tap1
+ exit 0
```

```
+ COMMAND=delete
+ COUNT=''
+ DEFCOUNT=2
+ [ -z lo ]
+ [ -z delete ]
+ [ delete = create ]
+ [ delete = delete ]
+ sudo sysctl net.link.tap.user_open=0
net.link.tap.user_open: 1 -> 0
+ sudo kldunload if_tap
+ sudo kldunload if_bridge
+ exit 0
```

**Comment by authmillenon** [comment]
Can you explain to me, what this output means?

**Comment by authmillenon** [comment]
adapted doc, addressed comments and (hopefully) fixed for OSX and FreeBSD.

**Comment by authmillenon** [comment]
@LudwigOrtmann ping?

**Comment by OlegHahm** [comment]
@LudwigOrtmann won't be avaliable until July 4.

**Comment by haukepetersen** [comment]
When @authmillenon and me just talked about this PR, it came to my mind if it makes sense to move this script out of the `cpu/native` location to somewhere more 'toolish', say `dist/tools/tapsetup/` or similar. What do you guys think about this?

**Comment by OlegHahm** [comment]
:+1:

**Comment by authmillenon** [comment]
> @LudwigOrtmann won't be avaliable until July 4.

Can someone else test this script then on FreeBSD and OSX?

File: cpu/native/tapsetup [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+        *)
+            ;;
+    esac
+}
+
+begin_tap() {
+    case "${PLATFORM}" in
+        FreeBSD)
+            sudo kldload if_tap || exit 1;;
+            sudo sysctl net.link.tap.user_open=1
```
Comment (cgundogan) [inline cpu/native/tapsetup]: my zsh is complaining here. Please move the `;;` to the end of this case

File: cpu/native/tapsetup [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+}
+
+create_tap() {
+    case "${PLATFORM}" in
+        FreeBSD)
+            echo "creating ${TAPNAME}${N}" || exit 1
+            sudo ifconfig tap${N} create || exit 1
+            sudo chown ${USER} /dev/tap${N} || exit 1
+            sudo ifconfig ${BRNAME} addm tap${N} || exit 1 ;;
+            sudo ifconfig tap${N} up || exit 1
```
Comment (cgundogan) [inline cpu/native/tapsetup]: here again, `;;` needes to be at the end

File: cpu/native/tapsetup (hunk 7 above)
Comment (OlegHahm) [inline cpu/native/tapsetup]: `/bin/sh` is a symlink to zsh on your system?

Comment (cgundogan) [inline cpu/native/tapsetup]: no it's not (: But still, this script refuses to work without these fixes

**Comment by authmillenon** [comment]
Addressed comments and fixed doc accordingly.

**Comment by authmillenon** [comment]
@thomaseichinger and @emmanuelsearch any luck testing this on OSX? @phiros any luck testing this on FreeBSD?

**Comment by emmanuelsearch** [comment]
I tested this on OS X, and it works as well as with #3290 (i.e. it sets up something, but ping6 timeouts 100% of the time in both directions). But I guess this PR was not focused on this issue. Basically: I think I can't test this more right now?

**Comment by authmillenon** [comment]
Since I don't have access to a OSX machine a more technical test would be far more helpful. This means: use tests/dev_eth with txtsnd instead of ping and a clarification were the packet exactly gets lost.  (Can you sniff the packets using wireshark or similar software? )

**Comment by authmillenon** [comment]
Or do you observe the same behavior with the old version of the script too. In that case, you are right: this is not the scope of this PR.

**Comment by LudwigOrtmann** [comment]
Works on FreeBSD 10

**Comment by authmillenon** [comment]
Can somebody ACK then?

**Comment by LudwigOrtmann** [comment]
AcK

**Comment by authmillenon** [comment]
Rebased and squashed

**Comment by authmillenon** [comment]
And finally merged.

**Pull request merged**